tokio-util/src/codec/framed_impl.rs

```diff
@@ -122,13 +122,20 @@ where
         let state: &mut ReadFrame = pinned.state.borrow_mut();
         loop {
             // Repeatedly call `decode` or `decode_eof` as long as it is
-            // "readable". Readable is defined as not having returned `None`. If
-            // the upstream has returned EOF, and the decoder is no longer
-            // readable, it can be assumed that the decoder will never become
-            // readable again, at which point the stream is terminated.
+            // "readable". Readable is defined as not having returned `None`.
+            // If the upstream has returned EOF, and the decoder is no longer
+            // readable, it might become readable again if the underlying
+            // AsyncRead is a FIFO or file.
+            // However, we still want to make sure that upon encountering an EOF
+            // we actually finish emmiting all it's associated decode_eof frames.
+            // Furthermore, we don't want to emit any decode_eof frames on retried
+            // reads after an EOF unless we've actually read more data.
             if state.is_readable {
                 if state.eof {
                     let frame = pinned.codec.decode_eof(&mut state.buffer)?;
+                    if frame.is_none() {
+                        state.is_readable = false;
+                    }
                     return Poll::Ready(frame.map(Ok));
                 }
```

Comment:
Are you sure, it seems like it is not possible for that to happen, since the `pausing → paused` transition always returns `Poll::Ready(None)` here?

Reply:
I'm not sure what "it" is referring to 🙇 so I'm gonna take a guess 😅. Setting `is_readable = false` is done because the state machine is operating across the boundaries of `poll_next` calls and therefore `loop` iterations.
So if a caller were to poll the stream after an eof it would hit the `decode_eof` call again (which would still return `None`) and thus never be able to restart consumption by actually reading to the buffer.

Reply:
The comments on line 161 and 162 are actually referring to functionality implemented on lines 198-200, which essentially corresponds to having the explicit `stopped` state in addition to `stopping`.

Reply:
I think it's worth to reiterate that the previous fusing behaviour is replaced by the ability to "restart" the stream even after it has returned `None`.
This is due to the fact that there are essentially two classes of underlying byte sources.
Non-resumables like pipes or sockets where eof actually means termination of the data stream, and resumables like FIFOs and (log) files.

I found the [wording of stream behaviour on returning `None`](https://docs.rs/futures-core/0.3.8/futures_core/stream/trait.Stream.html#panics) to be vague enough, that it should allow for restarting on these "weird" data sources.

The [documentation of `Read.read`](https://doc.rust-lang.org/nightly/std/io/trait.Read.html#tymethod.read), that `AsyncRead` is modeled after, is even more explicit on this:
> This reader has reached its "end of file" and will likely no longer be able to produce bytes. Note that this does not mean that the reader will always no longer be able to produce bytes.

If one takes the "should" in
> Poll::Ready(None) means that the stream has terminated, and poll_next should not be invoked again

as a "must", then this solution does not work.

A fix could be to:
- Either return `Poll::Pending` on a "paused" state, but that would mean that  a.) the stream will never exhaust on an EOF for files, and b.) the implementation would be enormous and brittle due to the non portability of filesystem notification mechanisms.
- Have FramedStream actually return `Poll<Option<Option<Self::Item>>>`, where the inner `Option` indicates the eof, while the outer `Option` returning `None` based on wether the underlying resource is actually resumable, which I don't think is easy to do or portable 🤔. It would also be a lot more backwards incompatible.

So I think the "proper/clean" solution is still biting the bullet, using the leeway in the wording of `Stream`, and having a stream that returns `None` on eof, but that can still be forced to restart by repeated calls to `poll_next`, relying on an external notification mechanism to do so, because it's always a safe operation. Additionally the "fuse on the first eof" behaviour can be regained by `fuse` 😄.

Reply:
Returning more items after `None` is fine. I just wanted to make sure we had a `None` between each restart of the stream.

Reply:
Do you think moving the state from the flags to an enum would make the code clearer?
I'm not experienced enough to gauge the performance, and maintainability implications of an enum, though.

Reply:
I've made some changes in an attempt to make the state transitions and states more apparent in the code.
I'm afraid this is as good as it's gonna get without an enum, and I hope that it actually improves
the situation instead of making it more confusing 😅